core/src/main/scala/org/apache/spark/rpc/netty/Inbox.scala

```diff
@@ -0,0 +1,220 @@
+    messages.add(OnStart)
+  }
+
+  /**
+   * Process stored messages.
+   */
+  def process(dispatcher: Dispatcher): Unit = {
+    var message: InboxMessage = null
+    synchronized {
+      if (!enableConcurrent && workerCount != 0) {
+        return
+      }
+      message = messages.poll()
+      if (message != null) {
+        workerCount += 1
+      } else {
+        return
+      }
+    }
+    while (true) {
+      safelyCall(endpoint) {
+        message match {
+          case ContentMessage(_sender, content, needReply, context) =>
+            val pf: PartialFunction[Any, Unit] =
+              if (needReply) {
+                endpoint.receiveAndReply(context)
+              } else {
+                endpoint.receive
+              }
+            try {
+              pf.applyOrElse[Any, Unit](content, { msg =>
+                throw new SparkException(s"Unmatched message $message from ${_sender}")
+              })
+              if (!needReply) {
+                context.finish()
+              }
+            } catch {
+              case NonFatal(e) =>
+                if (needReply) {
+                  // If the sender asks a reply, we should send the error back to the sender
+                  context.sendFailure(e)
+                } else {
+                  context.finish()
+                  throw e
+                }
+            }
+
+          case OnStart => {
+            endpoint.onStart()
+            if (!endpoint.isInstanceOf[ThreadSafeRpcEndpoint]) {
+              synchronized {
+                enableConcurrent = true
+              }
+            }
+          }
+
+          case OnStop =>
+            assert(synchronized { workerCount } == 1)
+            dispatcher.removeRpcEndpointRef(endpoint)
+            endpoint.onStop()
+            assert(isEmpty, "OnStop should be the last message")
+
+          case Associated(remoteAddress) =>
+            endpoint.onConnected(remoteAddress)
+
+          case Disassociated(remoteAddress) =>
+            endpoint.onDisconnected(remoteAddress)
+
+          case AssociationError(cause, remoteAddress) =>
+            endpoint.onNetworkError(cause, remoteAddress)
+        }
+      }
+
+      synchronized {
+        // "enableConcurrent" will be set to false after `onStop` is called, so we should check it
+        // every time.
+        if (!enableConcurrent && workerCount != 1) {
+          // If we are not the only one worker, exit
+          workerCount -= 1
+          return
+        }
+        message = messages.poll()
+        if (message == null) {
+          workerCount -= 1
+          return
+        }
+      }
+    }
+  }
+
+  def post(message: InboxMessage): Unit = {
+    val dropped =
+      synchronized {
+        if (stopped) {
+          // We already put "OnStop" into "messages", so we should drop further messages
+          true
+        } else {
+          messages.add(message)
+          false
+        }
+      }
+    if (dropped) {
+      onDrop()
```

Comment:
This is because onDrop accepts a `any` parameter, so Scala compiler fills `()` automatically here... I changed the parameter type to `InboxMessage` to fix it.